slow-query-analysis.md

```diff
@@ -0,0 +1,205 @@
+---
+title: 慢查询分析
+category: how-to
+---
+
+# 慢查询分析
+处理慢查询分为两步：
+1. 从大量查询中定位出哪一类查询比较慢；
+2. 分析这类慢查询的原因；
+
+第一步已经可以通过 [慢日志](/identify-slow-queries.md), [statement-summary](/statement-summary-tables.md) 等功能方便的定位，本文聚焦第二步。
+
+首先将慢查询归因成两大类：
+1. 优化器问题：如选错索引，选错 Join 类型或顺序；
+2. 系统性问题：将非优化器问题都归结于此类，如：某个 TiKV 实例忙导致处理请求慢，region 信息过期导致查询变慢；
+
+实际中，优化器问题可能造成系统性问题，如某类查询，优化器应使用索引，但却使用了全表扫，这可能导致这类 SQL 消耗大量资源，造成某些 KV 实例 CPU 飚高等，表现上看就是一个系统性问题，但本质是优化器问题。
+
+分析优化器问题需要有判断执行计划是否合理的能力，而系统性问题的定位相对简单，因此面对慢查询推荐的分析过程如下：
+1. 定位查询瓶颈：即查询过程中耗时多的部分；
+2. 分析系统性问题：根据瓶颈点，结合当时的监控/日志等信息，分析可能的原因；
+3. 分析优化器问题：分析是否有更好的执行计划；
+
+接下来会分别介绍上面几点。
+
+## 定位查询瓶颈
+定位查询瓶颈需要对查询过程有一个大致理解，TiDB 处理查询过程的关键阶段都在 [performance-map](https://raw.githubusercontent.com/pingcap/tidb-map/master/maps/performance-map.png) 图中了。
+
+定位单个查询的耗时，目前有两个工具：
+1. [慢日志](/identify-slow-queries.md)
+2. [explain analyze 语句](/sql-statements/sql-statement-explain-analyze.md)
+
+这两个工具侧重不同：
+1. 慢日志记录了 SQL 从解析到返回，几乎所有阶段耗时，较为全面；
+2. explain analyze 可以拿到 SQL 实际执行中每个执行算子的耗时，对执行耗时有更细分的统计；
+
+总的来说，利用慢日志和 `explain analyze` 可以比较准确的定位查询的瓶颈点，帮助你判断这条 SQL 慢在哪个模块（TiDB/TiKV），慢在哪个阶段，下面会有一些例子。
+
+## 分析系统性问题
+对于系统性问题，我们根据执行阶段，分成三个大类：
+1. TiKV 处理慢：如 coprocessor 处理数据慢；
+2. TiDB 执行慢：主要指执行阶段，如某个 Join 算子处理数据慢；
+3. 其他关键阶段慢：如取时间戳慢；
+
+拿到一个慢查询，我们应该先根据已有信息判断大致是哪个大类，再具体分析。
+
+### TiKV 处理慢
+如果是 TiKV 处理慢可以很明显的通过 `explain analyze` 中看出来，如下面这个例子，可以看到 `StreamAgg_8` 和 `TableFullScan_15` 这俩 `tikv-task` 花费了 170ms，而 TiDB 部分的算子耗时，减去这 170ms 后，耗时占比非常小，说明瓶颈在 TiKV。
```

Comment:
For slow query and statements, should we offer instructions based on TiDB Dashboard UI instead of SQL system table or even logs?

Reply:
It seems these two things don't conflict with each other.
Instructions based on Dashboard UI are at a high level and easy to use for our users, while SQLs or logs are more complex but can offer more details.